src/server.cpp

```diff
@@ -1523,16 +1536,21 @@ void Server::SendBreath(u16 peer_id, u16 breath)
 	Send(&pkt);
 }
 
-void Server::SendAccessDenied(u16 peer_id, AccessDeniedCode reason, const std::string &custom_reason)
+void Server::SendAccessDenied(u16 peer_id, AccessDeniedCode reason,
+		const std::string &custom_reason, bool reconnect)
 {
 	DSTACK(__FUNCTION_NAME);
 
 	NetworkPacket pkt(TOCLIENT_ACCESS_DENIED, 1, peer_id);
-	pkt << (u8) reason;
-
-	if (reason == SERVER_ACCESSDENIED_CUSTOM_STRING) {
-		pkt << custom_reason;
-	}
+	// additional validation
+	if (reason == SERVER_ACCESSDENIED_CUSTOM_STRING ||
+			reason == SERVER_ACCESSDENIED_SHUTDOWN ||
+			reason == SERVER_ACCESSDENIED_CRASH)
+		pkt << (u8)reason << custom_reason << (u8)reconnect;
+	else if (reason >= SERVER_ACCESSDENIED_MAX)
+		throw new SerializationError("Invalid access denial code");
+	else
+		pkt << (u8)reason;
```

Comment:
If you use the assert then you can re-up this insertion to the begining of the code